src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,440 @@
+                    error = str(e)
+
+                # Populate known Tool Log attributes
+                tool_log = {
+                    "inputs": args_to_inputs(func, args, kwargs),
+                    "output": output,
+                    "error": error,
+                }
+
+                # Write the Tool Log to the Span on HL_LOG_OT_KEY
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HL_LOG_KEY,
+                    value=tool_log,
+                )
+
+                # Return the output of the decorated function
+                return output
+
+        func.file = File(  # type: ignore
+            path=path if path else func.__name__,
+            type="tool",
+            version=tool_kernel,
+            is_decorated=True,
+            callable=wrapper,
+        )
+
+        return wrapper
+
+    return decorator
+
+
+def _build_tool_kernel(
+    func: Callable,
+    attributes: Optional[dict[str, Optional[Any]]],
+    setup_values: Optional[dict[str, Optional[Any]]],
+    strict: bool,
+) -> ToolKernelRequestParams:
+    """Build ToolKernelRequest object from decorated function."""
+    kernel = ToolKernelRequestParams(
+        source_code=textwrap.dedent(
+            # Remove the tool decorator from source code
+            inspect.getsource(func).split("\n", maxsplit=1)[1]
+        ),
+        # Note: OTel complains about falsy values in attributes, so we use OT_EMPTY_ATTRIBUTE
+        function=_build_function_property(
+            func=func,
+            strict=strict,
+        ),
+    )
+    if attributes:
+        kernel["attributes"] = attributes
+    if setup_values:
+        kernel["setup_values"] = setup_values
+    return kernel
+
+
+def _build_function_property(func: Callable, strict: bool) -> ToolFunctionParams:
+    """Build `function` property inside ToolKernelRequest."""
+    tool_name = func.__name__
+    description = func.__doc__
+    if description is None:
+        description = ""
+    return ToolFunctionParams(
+        name=tool_name,
+        description=description,
+        parameters=_build_function_parameters_property(func),  # type: ignore
+        strict=strict,
+    )
+
+
+class _JSONSchemaFunctionParameters(TypedDict):
+    type: str
+    properties: dict[str, typing.Union[dict, list]]
+    required: list[str]
+    additionalProperties: Literal[False]
+
+
+def _build_function_parameters_property(func) -> _JSONSchemaFunctionParameters:
+    """Build `function.parameters` property inside ToolKernelRequest."""
+    properties: dict[str, Any] = {}
+    required: list[str] = []
+    signature = inspect.signature(func)
+
+    for parameter in signature.parameters.values():
+        if parameter.kind in (
+            inspect.Parameter.VAR_POSITIONAL,
+            inspect.Parameter.VAR_KEYWORD,
+        ):
+            raise ValueError(f"{func.__name__}: Varargs and kwargs are not supported by the @tool decorator")
```

Comment:
is `Varargs` the right name?
(I'd suggest being explicit about `*args` and `**kwargs`  and mentioning those verbatim; like python's docs do https://docs.python.org/3/library/inspect.html#inspect.Parameter.kind )

Reply:
ack, changed